8Context Alerts Filter and Management Threshold

tlenaic opened the pull request:
@JordanGS 
you will find two issues add in zap plugin 
management threshold  and context alert filter

you will see bellow an eg of XML file

<?xml version="1.0" encoding="UTF-8"?>
<filters>
	<filter>
		<ruleId>40018</ruleId>
		<newLevel>1</newLevel>
		<uri>
			<url>http://a.b.c.*</url>
			<parameter></parameter>
		</uri>
		<urlIsRegex>true</urlIsRegex>
		<enabled>true</enabled>	
	</filter>
</filters>

best regards
lenaic

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -33,6 +33,8 @@
 import java.util.ArrayList;
 import java.util.concurrent.TimeUnit;
 
+import hudson.model.*;
```
No wildcard imports, the previous was intentional.

```
import hudson.model.AbstractBuild;
import hudson.model.AbstractProject;
import hudson.model.BuildListener;
import hudson.model.Computer;
import hudson.model.Node;
```

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -69,6 +66,7 @@
  * @author Thilina Madhusanka
  * @author Johann Ollivier-Lapeyre
  * @author Ludovic Roucoux
+ * @author Lenaic Tchokogoue
```
Most recent contributor at the top of the list.

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -338,7 +337,8 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
             return false;
         }
-        return res;
+        //return res;
```
Excess comment?

JordanGS commented:
@tlenaic Don't do a new commit, just update your previous.

`git commit --amend`

That's all i can see off the top of my head. I'll review properly once i'm done work today.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/config.jelly:
```diff
@@ -318,4 +330,43 @@ SOFTWARE.
+              <td><f:textbox field="hSoftValue"/></td>
+            </tr>
+            <tr>
+              <td>Meduim</td>
```
`Medium`.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/config.jelly:
```diff
@@ -318,4 +330,43 @@ SOFTWARE.
+          </table>
+          <table width="100%">
+            <tr>
+              <td>Cummulative Threshold</td>
```
`Cumulative`.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-buildThresholds.html:
```diff
@@ -0,0 +1,5 @@
+Post Build Management Threshold, allows zap to define if a build is flagges as passed, failed or unstable.</br></br>
```
`flagged`.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-buildThresholds.html:
```diff
@@ -0,0 +1,5 @@
+Post Build Management Threshold, allows zap to define if a build is flagges as passed, failed or unstable.</br></br>
+<b>Threshold value</b> stand for the weight of each alert severity</br></br>
+<b>Soft value</b> is threshold of each alert  severity</br></br>
+<b>Culmulative value</b> is the threshold of the all build</br></br>
```
`Cumulative`.

tlenaic reviewed (commented):
changes done
thanks

thc202 commented:
Typos were not yet fixed, there are wildcard imports, ideally, it should be imported just what's needed.

I'll pull the changes and try them, to comment further.

tlenaic commented:
my bad, 
i don't know all rules about zap plugin development.
i correct wildcard imports, if needed or any advice for me or preferences about those issues, feel free to ask me changes or change them your self.

thanks

thc202 commented:
OK, the dev guidelines are at:
https://wiki.jenkins-ci.org/display/JENKINS/zap-plugin+Code+Style+Guidelines

JordanGS commented:
@tlenaic just the spelling mistakes @thc202 pointed out above

I'm currently testing it.

JordanGS commented:
@tlenaic 

A couple things

1. Context Alert Filters, i need a better description on what this is, what it does and more precisely how to use it and how to write the xml file it's asking for. All of this needs to be documented or new user's won't know what to do. Please provide an example so i can properly test this functionality.

**EDIT**: Referencing [Google Group Discussion ](https://groups.google.com/forum/#!topic/zaproxy-jenkins/c3XP3NTyHOI)to include example and explanation of Context Alert Filters

2. Define Build Management Threshold, just a question or so but how do you define what happens if the thresholds are not met or what we constitute as a pass/failure?

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -338,7 +342,7 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
             return false;
         }
-        return res;
+        return true;
```
Should this not be returning the boolean value of `Result res`?

tlenaic replied:
I think it should be the best way. I don't know, if we could get the boolean value of Result res.

JordanGS reviewed (changes requested):
Just that one question and the spelling mistakes @thc202 pointed out. Otherwise i'm running scans now and we'll see how the results compare to what i'm expecting, especially with context alert filter.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/config.jelly:
```diff
@@ -128,6 +128,9 @@ SOFTWARE.
               <f:textbox clazz="required" />
             </f:entry>
+            <f:entry title="${%Logged out Indicator}" field="loggedOutIndicator">
+              <f:textbox clazz="required" />
```
Is this field really required?

JordanGS replied:
@thc202 it should be an optional field, i don't use it myself.

tlenaic replied:
i added it because in a project i needed it.(because there were no logged in indicator)
but you can still build a project event if it is not defined. but you're write, i should have put it in an optional block

JordanGS replied:
@tlenaic the required is there to indicate that the project will fail if this value is not specified.

thc202 replied:
I'd suggest removing the required from both indicators, none of them is really required, also, from a use case mentioned in the mailing list only the logged out indicator would be set.

thc202 commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -1676,14 +1756,89 @@ private void setUpForcedUser(BuildListener listener, ClientApi clientApi, String
+                         *
+                         * @method addAlertFilter
+                         *
+                         * @param String apikey
```
This parameter does not exists any more.

tlenaic replied:
yes

tlenaic commented:
@JordanGS 

Define Build Management Threshold 
i have done what you explain me here .  https://issues.jenkins-ci.org/browse/JENKINS-39988

Actually the build is flagged as ABORTED if some thing goes wrong in the construction.

The build is  flagged as STABLE if you are under the soft and cumulative value. (you should be Strictly lower those value)

The build is UNSTABLE If you are above one of the threshold value.

What return the build can be changed. that depend on your appreciation and cases

An enhancement could be :
if we are under the soft and cumulative value the build is marked as pass,
if we are under one of them the build flagged as unstable 
and  finally if all is bad the build is goes failed.

i hope i answered your question, i will check on enhancement tomorrow and precise how works alerts filter

JordanGS commented:
@tlenaic The context alert filter is actually somewhat counter intuitive at the moment to how it should work from a UI point of view.

Alert Filter Feedback

1. This should not be it's own entity in the UI but rather a checkbox under **Session Properties**, because we're trying to minimize the UI as much as possible.
2. The only path we ask for is the path for **ZAP Settings**, everything else should be relative to that. This should not be a text field/box but rather a drop down list that searches the alert filters folder for .filter files. (Similar to how policies and policy files are accepted.)
3. Filters themselves should be xml format rather than space separated to bring it in line with future ZAP import/export functionality.

JordanGS commented:
@tlenaic I'm busy with work until tomorrow afternoon, I'll re-check then. :)

tlenaic commented:
@JordanGS   it's ok, i am still working on those issues.
i realised i have some bug fixed now. i will push a clean version soon.

tlenaic commented:
@JordanGS @thc202 
new commit with fixed bugs
and a doc on how to use context alert filter here
https://groups.google.com/forum/#!topic/zaproxy-jenkins/c3XP3NTyHOI

Cheers

JordanGS commented:
@thc202 where do we want alert filters nested? (i didn't see a folder generated for it, unless i missed it)

How does `C:\Users\<username>\OWASP ZAP\alerts\filters\` sound as the default directory? Anything else that's related to alerts in the future can go into that the `alerts` folder are well, or should it be top level like `policies`, so `C:\Users\<username>\OWASP ZAP\filters\` , thoughts?

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -338,7 +342,7 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
             listener.error(ExceptionUtils.getStackTrace(e));
             return false;
```
If you're returning `return res.completeBuild;` below, is this return false still needed? (I'm asking, I'm not sure).

tlenaic replied:
yes. because res is not initailized

JordanGS commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/config.jelly:
```diff
@@ -128,6 +135,13 @@ SOFTWARE.
               <f:textbox clazz="required" />
             </f:entry>
+            <f:entry>
+            <f:optionalBlock title="${%Logged out Indicator}" field="activeLoggedOutIndicator" inline="true">
```
No need for optional block, simply make it

             <f:entry title="${%Logged out Indicator}" field="loggedOutIndicator">
               <f:textbox />
             </f:entry>

The text box should also not make it required.

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -54,16 +59,17 @@
-import org.zaproxy.clientapi.core.ApiResponseSet;
-import org.zaproxy.clientapi.core.ClientApi;
-import org.zaproxy.clientapi.core.ClientApiException;
+import org.zaproxy.clientapi.core.*;
```
No wildcard imports allowed. `.*`, use all the individual imports needed

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -1609,7 +1690,7 @@ private String setUpUser(BuildListener listener, ClientApi clientApi, String con
      * @param clientApi
      *            of type ClientApi: the ZAP client API to call method.
-     * @param contextId
+     * @param contextid
```
Should be `contextId`, not `contextid`

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -2083,7 +2316,7 @@ public ZAPDriverDescriptorImpl() {
             load();
         }
-
+        
```
Remove excess tabbing/spacing please.

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -2131,7 +2364,7 @@ public FormValidation doCheckLoadSession(@QueryParameter("loadSession") final St
          * Todo
          * 
-         * @param includedURL
+         * @param includedURL    
```
Remove excess tabbing/spacing please.

JordanGS commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/config.jelly:
```diff
@@ -115,6 +115,13 @@ SOFTWARE.
 
     <f:entry>
+
+      <f:optionalBlock title="${%Context Alert Filters}" field="loadAlertsFilter" inline="true">
```
Rename to `Alert Filters` from `Context Alert Filters`. (it's implied) since it's now nested under the context configuration. :)

JordanGS commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-loadAlertsFilter.html:
```diff
@@ -0,0 +1,2 @@
+Context Alert Filters allow you to automatically override the risk levels of any alerts raised by the active and passive scan rules within a context.</br></br>
```
This entire help box should be changed to match similar boxes.

Example: 

```
Checkbox to signify if the Alert Filters will be <b>applied(checked)</b> or <b>not applied(unchecked)</b> to the attack mode results.`</b>.</br></br><hr/></br>
Alert Filters will override the risk levels of any specified alerts raised by the passive and active scanners within the given context.</br></br>
Please see <a href="https://github.com/zaproxy/zaproxy/wiki/FAQconfig">HelpAddonsAlertFiltersAlertFilter</a> for more information.
```

JordanGS commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-xmlAlertsFilter.html:
```diff
@@ -0,0 +1 @@
+Select an XML Alert Filter File to use in the Context. The file must be stored in the Path provided for 'ZAP Settings'. </br></br><hr/></br>
```
Change to:

`Select a ZAP alert filter to use in the context. The file must be stored in the Path provided for 'ZAP Settings'.`

Changed the wording, notice that i removed the excess `</br></br><hr/></br>`

JordanGS reviewed (changes requested):
@tlenaic 

For now, the default directory for Alert Filters will be

Directory: `C:\Users\<username>\OWASP ZAP\alertfilters`

Filetype: and the extension for each file will be `.alertfilter`

The files themselves should be in xml format but they will have their own unique extension.

--

That's everything for logged out indicator and alert filters, once those changes are done. Those two can be pushed in.

Looks great! Once those changes are done, i'll take a look at threshold.

tlenaic commented on src/main/java/org/jenkinsci/plugins/zap/ZAPBuilder.java:
```diff
@@ -338,7 +342,7 @@ public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListen
             listener.error(ExceptionUtils.getStackTrace(e));
             return false;
```
yes. because res is not initailized

tlenaic commented:
@JordanGS 
Update done, let me know if something else.

JordanGS commented:
@tlenaic Okay so I've done a couple of things.

1. I've squashed all your commits into 1 and created a single commit message
2. I've taken the liberty of fixing some of the ordering on items, you changed it and it created a diff saying this line used to be on line 41. and now it's on 32., etc. I've undone those since they cause unnecessary noise in change history.
3. I've removed the required from the logged out indicator, since it's an optional field. Although i do understand that you need it :)

Overall this makes it easier to review and test. See notes below for Alert Filter. I'll try to find some time tomorrow or Friday to test thresholds and it should be released next week :)

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-loadAlertsFilter.html:
```diff
@@ -0,0 +1,3 @@
+Checkbox to signify if the Alert Filters will be <b>applied(checked)</b> or <b>not applied(unchecked)</b> to the attack mode results.`</b>.</br></br><hr/></br>
+Alert Filters will override the risk levels of any specified alerts raised by the passive and active scanners within the given context.</br></br>
+Please see <a href="https://github.com/zaproxy/zaproxy/wiki/FAQconfig">HelpAddonsAlertFiltersAlertFilter</a> for more information.
```
The link/text does not seem to be correct.

thc202 commented on src/main/resources/org/jenkinsci/plugins/zap/ZAPDriver/help-buildThresholds.html:
```diff
@@ -0,0 +1,6 @@
+Post Build Management Threshold, allows zap to define if a build is flagged as passed, failed or unstable.</br></br>
+<b>Threshold value</b> stand for the weight of each alert severity</br></br>
+<b>Soft value</b> is threshold of each alert  severity</br></br>
+<b>Cumulative value</b> is the threshold of the all build</br></br>
```
"of the all build" -> "of all the build" ?

tlenaic replied:
no, "of the build" or "of the current build". i will change it

JordanGS commented:
@tlenaic Error handling question for Alert FIlters

If the checkbox is selected but nothing is chosen from the drop down list, what happens?

Also, the functionality seems to have broken when using an environment variable for the Settings Path.

ERROR: java.io.FileNotFoundException: C:\Users\Goran\${ZAPROXY_SETTINGS}\alertfilters (The system cannot find the path specified)

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -1676,14 +1756,88 @@ private void setUpForcedUser(BuildListener listener, ClientApi clientApi, String
+        Utils.lineBreak(listener);
+        Utils.loggerMessage(listener, 0, "START : PARSING XML ALERT(s) FILTER File");
+
+        Path pathAuthScriptsDir = Paths.get(zapSettingsDir,NAME_ALERT_FILTERS_DIR_ZAP);
```
Should be 
`Path pathAuthScriptsDir = Paths.get(evaluatedZapSettingsDir, NAME_ALERT_FILTERS_DIR_ZAP);`

Or else it will break when using environment variables.

Also the variable name `pathAuthScriptsDir` doesn't make sense (i assume this is from when you copy-pasted). Please change it to `pathAlertFiltersDir`

FINAL: `Path pathAlertFiltersDir = Paths.get(evaluatedZapSettingsDir, NAME_ALERT_FILTERS_DIR_ZAP);`

Any variables created, should be named to explain their function/purpose.

JordanGS reviewed (changes requested):
@tlenaic Error handling question for Alert FIlters

If the alert filter checkbox is selected but nothing is chosen from the drop down list, what happens? What should happen? Proper work flow would be that if the checkbox is selected but nothing is chosen from the dropdown list, then an error would be thrown and ZAP should not execute. This check would be performed in the `checkParams` step/method.

Because right now if nothing is selected, you're letting an error get thrown without catching it.

```
START : PARSING XML ALERT(s) FILTER File
ERROR: java.io.FileNotFoundException: C:\Users\Goran\OWASP ZAP\alertfilters (The system cannot find the file specified)
```

Add a check under: `PLUGIN VALIDATION (PLG), VARIABLE VALIDATION AND ENVIRONMENT INJECTOR EXPANSION (EXP)"`

Example: 
```
[ZAP Jenkins Plugin] PLUGIN VALIDATION (PLG), VARIABLE VALIDATION AND ENVIRONMENT INJECTOR EXPANSION (EXP)
	ZAP PATH = [ C:\Development\OWASP\Zed Attack Proxy\2.6.0 ]
	(EXP) HOST = [ 127.0.0.1 ]
	(EXP) PORT = [ 9090 ]
	(EXP) LOAD SESSION = [ C:\Development\Jenkins\current\workspace\tlenaic_test\goran.session 
C:\Development\Jenkins\current\workspace\tlenaic_test\goran.session ]
	(EXP) ZAP SETTINGS DIRECTORY = [ C:\Users\Goran\OWASP ZAP ]
	(EXP) CONTEXT NAME = [ Context Tlenaic 5 ]
	(EXP) INCLUDE IN CONTEXT = [ www.abc.com ]
	(EXP) EXCLUDE FROM CONTEXT = [  ]
	(EXP) STARTING POINT (URL) = [ www.abc.com ]
	ALERT FILTERS = [ TRUE ]
	ALERT FILTERS FILE = [ C:\Users\Goran\OWASP ZAP\alertfilters\list.alertfilters ]
```

Notice the two new lines at the bottom.

So if **ALERT FILTERS** is true but **ALERT FILTERS FILE** is null or empty string, then `throw new IllegalArgumentException`.

JordanGS commented:
@tlenaic let me know if you have any questions or concerns :)

JordanGS commented:
@tlenaic stop making changes for the moment to `evaluatedZapSettings`, there may be complications, speaking to Jenkins developers.

I found a bug last night.

tlenaic commented:
@JordanGS ok thanks, however i have done all changes you've ask for

JordanGS commented on src/main/java/org/jenkinsci/plugins/zap/ZAPDriver.java:
```diff
@@ -1676,14 +1763,88 @@ private void setUpForcedUser(BuildListener listener, ClientApi clientApi, String
+        Utils.lineBreak(listener);
+        Utils.loggerMessage(listener, 0, "START : PARSING XML ALERT(s) FILTER File");
+
+        Path pathAlertFiltersDir  = Paths.get(evaluatedZapSettingsDir,NAME_ALERT_FILTERS_DIR_ZAP);
```
Change this to `zapSettingsDir` please, sorry. I found out that the config page cannot evaluate environment variables from the slave. So I'll be removing the evaluated value for `zapSettingsDir`

`Path pathAlertFiltersDir  = Paths.get(zapSettingsDir, NAME_ALERT_FILTERS_DIR_ZAP);`

^notice the space after the ,

Reference: https://groups.google.com/forum/#!topic/jenkinsci-dev/_SC7dwWpIb8

Sorry about that, i didn't know this was an issue until last night. I'll have to fix issues as well from other sides since this causes a bug in the system.

P.S. I have to go to work for now, i'll take a look again later tonight when i can.

tlenaic replied:
changed. Ok, see you

JordanGS committed:
New Feature Implementation for v1.1.0

-implemented new feature Alert Filters (JENKINS-39985)
-implemented new feature Threshold Management (JENKINS-39988)
-added logged out indicator to authentication
-minor bug fixes / code clean up

Update help-buildThresholds.html
check params for alerts filter
name of the path changed in alerts filter function
evaluatedZapSettingsDir to zapSettingsDir

JordanGS commented:
@tlenaic squashed commits, beginning review. Made some quick changes. I'll finish checking over tonight. I started a scan.

JordanGS committed:
Patch

-fixed console messages
-updated variable calls to include this modifier

JordanGS commented:
@tlenaic I'm making some minor changes, based on a discussion with @thc202 

I want to push out Alert Filters and merge it, would you mind if we separate the pull requests, so use this one for Alert Filters and we can do Thresholds in another PR and a 3rd for Logged out indicator? Or would you prefer to do it in the same PR?

Let me know before you make any changes thought please, currently updating.

tlenaic commented:
@JordanGS @thc202  if i have to choose i will prefer we do it in the same.
It will be easier, than separed them, push one feature, review thresholds and then push again. 
If reviews are finished for alert filters, why not, do the same for threshold this week (i am avaible) and push them for once.
Whatever, i'll do what you want, you know more the plugin, if it is better to, let's separeted them.
i dont care, if the two feature are provided now or later, once they are it's ok

JordanGS committed:
Bug Fixes / Changes for Alert Filters

-manual Indicated bracket rather than letting system handle it, causes issues when Windows/Linux
-updated variable names associated to alert filters to be more descriptive
-xml File selection was not being saved when going back into the config to update settings
-removed the checkbox to minimize the UI, Empty means None
-updated all Console messages
-added debug console messages
-removed extension from name in drop down list
-moved alert filter call above authentication (visually looks better) and update the code to reflect the change in order
-updated help files

JordanGS commented:
@tlenaic I've pushed a new version with various bug fixes and some minor changes. Those include the following

**Changes**
1. I removed the checkbox since there was only one element for it to hide, It actually created more vertical scrolling as a result. If in the future we have more elements for alert filters then we can use a checkbox. Let me know if you have any concerns, i'm fine either way on this one actually. The lack of a checkbox now implies `None` so no error will be thrown.

2. With the removal of the checkbox, i changed the variable name to `alertFilters` for simplicity. Also updated the associated help files.

3. Various variable names were changed to be more descriptive, especially for the XML part. This way it's easier and clearer for people to understand what's going on.

4. Moved Alert Filter Drop Down List above authentication checkbox since it visually looks better that way and update the code to reflect the change in order.

**Bug Fixes**

1. `String pathXmlFiltersFiles = pathAlertFiltersDir + "/" + xmlAlertsFilter;` is a very bad way of creating a path since Windows uses `\` and Linux uses `/`, this will cause issues depending on the operating system. Take a look at the change i made there. Windows 10 is a special case, it doesn't care if it's `\` or `/`.

2. When in the Job Config Page do the following. Select an Alert Filters File > Save > Go back into Job Config and the drop down would not have the specified file selected anymore. The data bound constructor was missing the required get method. Has been fixed.

3. Console messages had some upper case, some lower case letters. Everything was made upper case and brought more inline with how other messages were done. Additional messages were added to be able to properly debug for other users.

4. Visual change only, removed the extension from the drop down list, there is only one type of file we accept and in this case they don't need to see it (same as policy).

--

What i would like to ask you to do.

1. Check over the changes, test them as you see fit. Let me know if you have any questions or concerns? Or you're displeased with any of the changes. Please talk to me :)

2. If everything is good, please do **NOT** close this PR. But do the following.

- [ ] Create a PR for Logged Out Indicator
- [ ] Create a PR for Alert Filters
- [ ] Create a PR for Threshold Management

Logged out indicator and Alert Filters are done (once you sign off) and can be pushed into the jenkinsci and released. Unless you can find any other bugs with either?

So 3 separate Pull Requests.

JordanGS commented:
@tlenaic In regards to the PR, please do NOT close this one but do create separate PRs. I've spoken to @thc202 and we both agree that it will make version management easier. Sorry for thte trouble, this is my first time managing a git repository. I should have mentioned this sooner. I'm learning too :)

I will take a look at Threshold Management the same way i did Alert Filters as soon as i can (probably this weekend). I appreciate all your hard work on both :)

P.S. @thc202 will be adding import/export functionality to the ZAP GUI to make the Alert Filters functionality easier, that would not be possible without you. Thank you!

JordanGS commented:
@tlenaic To clarify tho, i am tracing a serialization error that was created in this commit which is preventing changes to any older jelly config elements. I believe the error is coming from alert thresholds. But that is also the need for separate PR's, it will be able to trace errors easier through versions.

For now I'm not sure what causes it, i'm testing 1.0.8 to try and replicate the issue. Will update tomorrow.

tlenaic commented:
@JordanGS @thc202 do we agree we have one alert filters with rules redefined ?
on the content.
with my algorithm and the loop on uri it will be more readable and project standpoint optimize. but in the other hand may be it is not maintenable and we have problems with GUI compliance

i join an explaination to the comment.
![alertfilters](https://cloud.githubusercontent.com/assets/22395028/24897521/7f288348-1e99-11e7-90bb-aca2881cd164.png)

tlenaic commented:
@JordanGS @thc202 thanks to you, for helping me with bugs and enhancement of those features

JordanGS commented:
@tlenaic the problem is that it's counter to how the GUI functionality works and as @thc202 explained to me it doesn't work that way. For example, say you have the following

`abc.com/main.html `(this one has an SQL Injection Error - **False Positive**)
`and abc.com/contact.html `(this one has an SQL Injection Error - but i want to lower the severity so it's not High, so **Info**)

Your proposed changes group all uri into a single alert filter, but the purpose of alert filters is on a per uri basis. That's why i had changed the structure to match the GUI functionality. Your algorithm is more optimized but it doesn't match the functionality of the GUI. So i don't think we should change the xml structure. @thc202 thoughts?

JordanGS commented:
@tlenaic let me know when you have a chance to create a new PR for Threshold Management, we'll review it from there together and we can close this one then.

tlenaic commented:
@JordanGS ok it will be done today

JordanGS commented:
@tlenaic please do all PR's against branch development-1.1.0 You have some changes to pull in via git.
